src/lib/shadow-shim-helper-rs/src/signals.rs

```diff
@@ -201,6 +201,21 @@ pub struct shd_kernel_sigaction {
     ksa_mask: shd_kernel_sigset_t,
 }
 
+impl shd_kernel_sigaction {
+    pub fn handler(&self) -> signal::SigHandler {
+        let handler_int: usize = unsafe { std::mem::transmute(self.u.ksa_handler) };
```

Comment:
I think the `transmute` is safe, but I'm happy to remove it :)

The `transmute` won't compile if the source and destination types aren't the same sizes, and Rust guarantees that the Option is zero-size in this case (https://doc.rust-lang.org/std/option/index.html#representation). We depend on this in the definition of `ShdKernelSigactionUnion`, which has to be bitwise-compatible with the corresponding C type.

Since there are no "extra" bits to store a validity flag, initalizing to `None` must initialize all bits to 0, so there should never be any "undefined bits".

But yeah, your version is definitely easier to understand and be confident of its safety :)

Reply:
I agree in theory, but am always worried about what assumptions rust makes :) The transmute documentation has this:

> Note that using transmute to turn a pointer to a usize is (as noted above) undefined behavior in const contexts. Also outside of consts, this operation might not behave as expected – this is touching on many unspecified aspects of the Rust memory model. Depending on what the code is doing, the following alternatives are preferable to pointer-to-integer transmutation: [...]